src/site/content/en/payments/orchestrating-payment-transactions-with-a-service-worker/index.md

```diff
@@ -0,0 +1,478 @@
+Use this to construct a UI to let the customer select a shipping option.
+        </td>
+      </tr>
+    </tbody>
+  </table>
+</div>
+
+## Open the payment handler window to display the web-based payment app frontend
+
+When a `paymentrequest` event is received, the payment app can open a payment
+handler window by calling `PaymentRequestEvent.openWindow()`. The payment
+handler window will present the customers your payment app's interface where
+they can authenticate, choose shipping address and options, and authorize the
+payment. We'll cover how to write the frontend code in [Handling payments on the
+payment
+frontend](https://docs.google.com/document/d/18-zDc1k0s_9529S6TJi3dFT86cxXMTo044Ehk7VjRds/edit#).
+
+<figure class="w-figure" style="width:300px; margin:auto;">
+  <video controls autoplay loop muted class="w-screenshot">
+    <source src="https://storage.googleapis.com/web-dev-assets/payments/web-based-payment-app.webm" type="video/webm">
+    <source src="https://storage.googleapis.com/web-dev-assets/payments/web-based-payment-app.mp4" type="video/mp4">
+  </video>
+  <figcaption class="w-figcaption">
+    Checkout flow with a web-based payment app.
+  </figcaption>
+</figure>
+
+Pass a preserved promise to `PaymentRequestEvent.respondWith()` so that you can
+resolve it with a payment result in the future.
+
+{% Label %}
+[payment handler] serviceworker.js:
+{% endLabel %}
+
+```js
+…
+self.addEventListener('paymentrequest', async e => {
+…
+  // Retain a promise for future resolution
+  // Polyfill for PromiseResolver is provided below.
+  resolver = new PromiseResolver();
+
+  // Pass a promise that resolves when payment is done.
+  e.respondWith(resolver.promise);
+  // Open the checkout page.
+  try {
+    // Open the window and preserve the client
+    client = await e.openWindow(checkoutURL);
+    if (!client) {
+      // Reject if the window fails to open
+      throw 'Failed to open window';
+    }
+  } catch (err) {
+    // Reject the promise on failure
+    resolver.reject(err);
+  };
+});
+…
+```
+
+{% Aside %}
+Use a convenient `PromiseResolver` polyfill to resolve a promise at arbitrary timing.
+```js
+class PromiseResolver {
+  constructor() {
+    this.promise_ = new Promise((resolve, reject) => {
+      this.resolve_ = resolve;
+      this.reject_ = reject;
+    })
+  }
+  get promise() { return this.promise_ } 
+  get resolve() { return this.resolve_ }
+  get reject() { return this.reject_ }
+}
+```
+{% endAside %}
+
+{% Aside %}
+For security reasons, the frontend page opened in the payment handler window
+must have valid HTTPS certificates and no [mixed
+content](https://developers.google.com/web/fundamentals/security/prevent-mixed-content/what-is-mixed-content);
+otherwise the payment request will be cancelled by Chrome. Learn more at
+[Debugging a web-based payment
+app](https://docs.google.com/document/d/1vfgHeuvskPlE18l_IXtlijP3hGomC4FT52b4MqS4BiM/edit#heading=h.genktfd5wi5v).
```

Comment:
This links to a doc that has `[Ready for reviews]` in the title. It's OK if we need to link to a Doc (although it seems like we should eventually get this content into your collection) but I think we need to link to a more polished Doc

![image](https://user-images.githubusercontent.com/4713486/89962023-602aa000-dbf8-11ea-9765-83853f7170f5.png)